feat(extui): show "g<" hint when message spills 'cmdheight'

luukvbaal opened the pull request:
**feat(extui): show "g<" hint when message spills 'cmdheight'**

Problem:  Extui shows a spill indicator to hint to the user to press "g<"
          to show the output of the last command. The user may be
          unaware of this mapping.
Solution: Route a hint message to the message "box" window.

**fix(extui): adjust which messages are sent to "more" window**

Problem:  Decision whether message is sent to "more" window is based on
          the number of newlines present in a message, rather than the
          actual text height.
          With the "box" target, messages that come from a cmdline
          entered command are not always routed to the more window.
Solution: Still write the message to the target buffer, and calculate
          the actual text height. Postpone updating several state
          variables until after the decision to re-route is made.
          With the "box" target, only consider the text height of the
          message if it is not a message from a cmdline entered command.

Fix #33956

luukvbaal committed:
feat(extui): show "g<" hint when message spills 'cmdheight'

Problem:  Extui shows a spill indicator to hint to the user to press "g<"
          to show the output of the last command. The user may be
          unaware of this mapping.
Solution: Route a hint message to the message "box" window.

brianhuster commented on runtime/lua/vim/_extui/messages.lua:
```diff
@@ -265,13 +265,16 @@ function M.show_msg(tar, content, replace_last, more)
+      local spill = M.cmd.lines - ext.cmdheight
+      M.virt.msg[M.virt.idx.spill][1] = spill > 0 and { 0, ('[+%d]'):format(spill) } or nil
+      if spill > 0 then
+        M.msg_show('verbose', { { 0, ('Press g< to show %d more lines'):format(spill), 0 } })
```
I think the verb "see" is more suitable than "show" here (the subject is "user", right?)

luukvbaal committed:
fix(extui): adjust which messages are sent to "more" window

Problem:  Decision whether message is sent to "more" window is based on
          the number of newlines present in a message, rather than the
          actual text height.
          With the "box" target, messages that come from a cmdline
          entered command are not always routed to the more window.
Solution: Still write the message to the target buffer, and calculate
          the actual text height. Postpone updating several state
          variables until after the decision to re-route is made.
          With the "box" target, only consider the text height of the
          message if it is not a message from a cmdline entered command.

brianhuster commented:
> inline the hint into the message (right-aligned).

This was what I thought when writing #33956

luukvbaal commented:
> This was what I thought when writing https://github.com/neovim/neovim/issues/33956

Miscommunication then, I thought you were specifically referring to the message 'box' window. I'd rather not append a long-form hint to the actual message area. That will just lead to more of the message being cut-off. An ephemeral hint message in the 'box' window seemed like the right fit to me but yeah maybe it is too intrusive when it is shown every time.

brianhuster commented:
So is it possible to show that message in the `cmdheight + 1`th line?

luukvbaal commented:
Reverted in #34003. What do you mean? The `cmdheight + 1th` line is not visible?

brianhuster commented:
> What do you mean? The cmdheight + 1th line is not visible?

Nevermind, because after updating Nvim just now, it always show and focus on the `msgmore` window whenever message exceeds `'cmdheight'`, even though I don't press `g<`. So now it behaves almost like `Press q to continue`

To reproduce:
1. `nvim --clean -c "require 'vim._extui'.enable {}"`
2. `:lua assert(1 == 2)`

luukvbaal commented:
> So now it behaves almost like Press q to continue now.

For interactively cmdline typed commands, yes: https://github.com/neovim/neovim/pull/33885#issuecomment-2863726493.
